### DISCUSSION TITLE
Listing missing local files (improved warning)

### DISCUSSION THREAD

**Discussion by gleicher** [marp-team/marp#257] [Q&A]
Is there a way to get a list of what local files are missing?

I am using marp cli to create PDFs. (this is a great workflow for my class lectures)

The terse error message is usually good. But sometimes, being told "[  WARN ] Some of the local files are missing and will be ignored. Make sure the file paths are correct." isn't enough - I'd like to know which local files (what lines in the markdown file).
(skimming over a 100 page lecture to find the missing image can be tricky)

Is there a way to get more verbose warning messages from the CLI? 

Thank you,

Mike

**Comment by yhatt** [comment]
Marp CLI does not have an ability to show missing files, but you can do the same thing easily by using the browser with HTML export.

Export HTML into next to the Markdown file and open it in the browser that has enabled Developer Tools (<kbd>F12</kbd>). In the "Network" tab, missing resources while loading HTML will highlight.

![スクリーンショット 2022-01-28 3 15 18](https://user-images.githubusercontent.com/3993388/151419444-68f21f73-7886-43c6-a7f4-6393019ebb67.png)

**Reply by gleicher** [reply]
Thanks!
That's a good tip. 
Unfortunately, what it tells me is that there isn't a missing file. (well, the browser can't find my fonts - which the PDF version is clearly getting right). This makes it myseterious why the pdf generation tells me there is a missing file. The resulting PDF looks good...

Maybe someday I will try to add a "verbose warnings" option that shows this (I am such a big fan of Marp that I'd like to find a way to contribute. Except that right now, I am so busy using it to prepare my class...)

**Reply by yhatt** [reply]
Marp CLI will convert HTML generated from Markdown to PDF so the above solution should be simulating to fetch local resources in the process of PDF conversion. However, sometimes may not reproduce exactly due to the difference of Chrome profiles (the profile for browsing web and the exclusive profile for Marp conversion). For example, that difference would be big if your organization was forcing specific configurations, extensions, or policies to Chrome. (`chrome://policy` to see the current policy) In addition, some of policies may have negative effects to Marp conversions.

`--verbose` option would become a good contribution to Marp CLI. We are always welcome to contribute :)

**Comment by Helle1** [comment]
I am having a similar problem. 
My --html option works fine (no missing files in the browser) and when I use --pdf it says I am missing a file.

specifically, it is about a css file. I try to @import a sub-css file in my theme file.

```css
/* @theme my-theme */

@import 'default'
@import '/themes/utils/sub-file.css'

...
```

my shell promt is:

```shell
$ npx marp slides/slideTemplate.md --theme themes/my-theme.css --allow-local-files --html

[  INFO ] Converting 1 markdown...
[  INFO ] slides/slideTemplate.md => slides/slideTemplate.html

$ npx marp slides/slideTemplate.md --theme themes/my-theme.css --allow-local-files --html --pdf

[  INFO ] Converting 1 markdown...
[  WARN ] Insecure local file accessing is enabled for conversion from slides/slideTemplate.md.
[  WARN ] The local file is missing and will be ignored. Make sure the file path is correct.
[  INFO ] slides/slideTemplate.md => slides/slideTemplate.pdf
```

The substyle is missing in the pdf, but working in html version. 

What did I get wrong with the include?

**Reply by yhatt** [reply]
_Marp team does not recommend importing external resources in CSS due to many troubles._

While the process of PDF conversion, [Marp CLI will generate an intermediate HTML file to OS-specific temp directory](https://github.com/marp-team/marp-cli/blob/eb6d82ca0369652e9743a284a893a3c4964626b2/src/converter.ts#L547-L550) (if allowed using local files). When generating, we instruct the correct base directory to HTML by using `<base>` tag. Resources in Markdown (e.g. images) respect this setting. However, the browser do not respect it for external resources specified in CSS.
https://stackoverflow.com/questions/2161377/is-the-html-base-tag-also-honored-by-scripting-and-css

If used relative path in style, the path always resolve from the directory that the HTML file is located. e.g. `/var/tmp` or `C:\Users\%USER%\AppData\Local\Temp` for PDF conversion.

Using absolute path, such as `https://xxx.yyy/zzz` (there is an absolute path determined by URI), might be temporarily resolved, but the created theme may become incompatible with other Marp tools because using `@import` for external CSS files will bypass Marpit's CSS scoping.
https://github.com/marp-team/marpit/issues/363#issuecomment-1636352542

Thus, the most stable solution is stop using `@import` for external CSS files. If you want to keep using relative path in `@import`, you also can try to inline external CSS contents with PostCSS and [postcss-import](https://github.com/postcss/postcss-import) or [postcss-url](https://www.npmjs.com/package/postcss-url) before using in Marp. ([Actually Marp Core is inlining `url()` on build time powered by postcss-url](https://github.com/marp-team/marp-core/blob/76aec583d703d21ac4dcaa8cce12008798543b81/rollup.config.mjs#L44-L48))

**Reply by Helle1** [reply]
Wow thanks for the quick and deep and complete answer. Thank you very much. 
I was already also thinking about the postcss import approach. I will go for it.

**Reply by Helle1** [reply]
My solution now is to preprocess the css with less in npm.
You need to add ignore the `@import (inline) "default";` directive with the `(inline)` word, and change the `theme.css` to `theme.less` as well as the sub-file.

my `themes.less`-file looks like this now:

```css
 /* @theme my-theme */

@import (inline) 'default'
@import 'utils/sub-file'

```

and my cli call:

```shell
$ npx less themes/my-theme.less > themes/my-theme.css && npx marp slides/slideTemplate.md --theme themes/my-theme.css --allow-local-files --html --pdf
```

that works fine (you need to install less: `npm i less --save-dev`)

**Comment by boborbt** [comment]
I see a similar problem: if I output to a pdf I get the ``missing files`` warning, but if I just change the output to html the warning is not displayed. In both cases the output looks ok.

My cmd lines are:

```bash
marp 05-Ensembles/ensembles.md --allow-local-files --html -o output/05-Ensembles.pdf --theme-set ./theme/themes
```
and
```bash
marp 05-Ensembles/ensembles.md --allow-local-files --html -o output/05-Ensembles.html --theme-set ./theme/themes
```

It is clearly not a big problem, but maybe there is some bug lurking in marp-cli?

**Reply by yhatt** [reply]
No warnings about missing files in HTML conversion are expected behavior, and there are no problems.

Warnings about invalid local file access only appear when the internal browser process is used for conversion, such as with PDF, PPTX, and images. This is because the rendering result can differ depending on whether the file assets can be resolved during conversion. If they cannot be resolved, warnings will be presented by CLI.

In contrast, HTML conversion does not use a browser internally. It simply converts Markdown text into HTML text. Whether the file assets can be resolved is determined only when the presentation is opened in the user's browser. Therefore, no warnings are shown during CLI conversion.

In HTML conversion, browser's dev tools would be helpful to check the error of assets, as shown in https://github.com/orgs/marp-team/discussions/257#discussioncomment-2061133

**Reply by boborbt** [reply]
Ok, makes sense. Thank for the prompt answer.

**Reply by boborbt** [reply]
May I add another issue I’ve noticed regarding the difference between marp-cli and the HTML rendering (specifically the VS Code preview, which I assume uses HTML)?

I’m attaching two screenshots of the same slide for comparison.

Rendered in VS Code:
<img width="1451" height="813" alt="rendered-in-the-vscode-plugin" src="https://github.com/user-attachments/assets/6316da07-ea02-470c-abb7-366df81a8c6f" />

Rendered by marp-cli (using the command I shared above):
<img width="1185" height="665" alt="rendered-by-marp-cli-pdf" src="https://github.com/user-attachments/assets/4ccbbd86-e189-4dee-9b36-9a3d48995b14" />

As you can see the page number is rendered as a white box in the pdf, while it was correctly displayed in the vscode rendering.

This makes me wonder if the issue is related to my custom theme (a heavily modified version of uncover, available [here](https://github.com/boborbt/marp-uncover-bb-theme)). I had assumed the PDF output should match the HTML rendering, but perhaps I’ve introduced something unusual in the theme that affects marp-cli rendering?

**Reply by yhatt** [reply]
If you're using Mac, it's known that PDF viewers such as Preview.app, Quick Look in Finder, and viewers using Mac's component do not render some PDF elements correctly.

If the output PDF looks good in other viewers that are not using Mac's PDF rendering, such as Google Chrome, Firefox, Adobe Acrobat, etc, there should be no problems in Marp and PDF, because it's just a bug in Mac PDF rendering.

- Related:
  - https://github.com/marp-team/marp-cli/issues/619#issuecomment-2466745819
  - https://issues.chromium.org/issues/41476907

**Reply by boborbt** [reply]
That was it, I checked in chrome and they are fine. Sorry for the misplaced blame then 😞.

Thanks once more for your amazing support.
